File: src/lib/optimizer.ts
import { extractKeywords, inferJobDetails } from "@/lib/keywords";
import { callOpenAIJson } from "@/lib/openai";
import type { CareerAdvice, MatchResult, OptimizationResult, ResumeRecord } from "@/lib/types";

const optimizationSchema = {
  type: "object" as const,
  additionalProperties: false,
  required: ["optimizedText", "summary", "addedKeywords", "improvedBullets", "jobTitle", "companyName"],
  properties: {
    optimizedText: { type: "string" },
    summary: { type: "string" },
    addedKeywords: {
      type: "array",
      items: { type: "string" }
    },
    improvedBullets: {
      type: "array",
      items: { type: "string" }
    },
    jobTitle: { type: "string" },
    companyName: { type: "string" }
  }
};

const adviceSchema = {
  type: "object" as const,
  additionalProperties: false,
  required: ["shouldApply", "reasoning", "growthPotentialScore", "skillGapAnalysis", "suggestedFocus"],
  properties: {
    shouldApply: { type: "boolean" },
    reasoning: { type: "string" },
    growthPotentialScore: { type: "number" },
    skillGapAnalysis: {
      type: "array",
      items: { type: "string" }
    },
    suggestedFocus: {
      type: "array",
      items: { type: "string" }
    }
  }
};

export async function optimizeResume(input: {
  resume: ResumeRecord;
  jd: string;
  match?: MatchResult;
  jobTitle?: string;
  companyName?: string;
}): Promise<OptimizationResult> {
  const jobDetails = inferJobDetails(input.jd);
  const jobTitle = input.jobTitle || jobDetails.jobTitle;
  const companyName = input.companyName || jobDetails.companyName;
  const missingKeywords =
    input.match?.missingKeywords.length ? input.match.missingKeywords : getMissingKeywords(input.jd, input.resume.text);

  const aiResult = await callOpenAIJson<OptimizationResult>({
    schemaName: "optimized_resume",
    schema: optimizationSchema,
    system: [
      "You are an expert ATS resume editor.",
      "Edit only resume text content while preserving the source resume's section order, voice, and formatting intent.",
      "Do not invent employers, credentials, degrees, dates, metrics, or tools.",
      "Add only truthful, JD-relevant keywords that are supported by the resume context.",
      "Keep the result concise enough for one page, around 450 to 650 words maximum.",
      "Return JSON only."
    ].join(" "),
    user: JSON.stringify({
      jobTitle,
      companyName,
      jobDescription: input.jd,
      currentResumeText: input.resume.text,
      missingKeywords,
      requiredRules: [
        "ATS-friendly text",
        "one page only",
        "preserve section order",
        "preserve bullet style as much as plain text allows",
        "no font or layout instructions in the text"
      ]
    })
  }).catch(() => null);

  if (aiResult?.optimizedText) {
    return {
      ...aiResult,
      optimizedText: trimToOnePage(aiResult.optimizedText),
      jobTitle,
      companyName
    };
  }

  return fallbackOptimization(input.resume.text, input.jd, missingKeywords, jobTitle, companyName);
}

export async function generateCareerAdvice(input: {
  jd: string;
  resume?: ResumeRecord | null;
  match?: MatchResult | null;
}): Promise<CareerAdvice> {
  const jdKeywords = extractKeywords(input.jd, 35);
  const resumeText = input.resume?.text ?? "";
  const resumeKeywords = extractKeywords(resumeText, 35);
  const missing = jdKeywords.filter((keyword) => !resumeKeywords.includes(keyword)).slice(0, 8);
  const score = input.match?.score ?? Math.max(35, 86 - missing.length * 7);

  const aiResult = await callOpenAIJson<CareerAdvice>({
    schemaName: "career_advice",
    schema: adviceSchema,
    system: [
      "You are a practical career advisor.",
      "Use the JD and available resume evidence to decide whether the user should apply.",
      "Be direct, realistic, and concise.",
      "Return JSON only."
    ].join(" "),
    user: JSON.stringify({
      jobDescription: input.jd,
      resumeText,
      matchScore: input.match?.score,
      missingKeywords: missing
    })
  }).catch(() => null);

  if (aiResult) {
    return {
      ...aiResult,
      growthPotentialScore: clampScore(aiResult.growthPotentialScore)
    };
  }

  return {
    shouldApply: score >= 58,
    reasoning:
      score >= 72
        ? "The resume already overlaps with the JD on core responsibilities and keywords."
        : score >= 58
          ? "The role is reachable with targeted resume optimization and a focused application."
          : "The JD shows several gaps that may weaken the application unless there is unstated experience.",
    growthPotentialScore: clampScore(score + Math.min(12, missing.length * 2)),
    skillGapAnalysis: missing.length ? missing.map((keyword) => `Make ${keyword} visible if you have real experience.`) : ["No major keyword gaps detected."],
    suggestedFocus: missing.slice(0, 4)
  };
}

function fallbackOptimization(
  resumeText: string,
  jd: string,
  missingKeywords: string[],
  jobTitle: string,
  companyName: string
): OptimizationResult {
  const lines = resumeText
    .split(/\r?\n/)
    .map((line) => line.trimEnd())
    .filter((line) => line.trim().length > 0);

  const keywordsToAdd = missingKeywords.slice(0, 8);
  const improvedLines: string[] = [];
  let addedKeywordIndex = 0;
  const bulletPrefix = /^(\s*[-*•]\s+)/;

  for (const line of lines) {
    const match = line.match(bulletPrefix);
    if (match && addedKeywordIndex < keywordsToAdd.length && line.length < 180) {
      const keyword = keywordsToAdd[addedKeywordIndex];
      addedKeywordIndex += 1;
      improvedLines.push(`${match[1]}${strengthenBullet(line.replace(bulletPrefix, ""), keyword)}`);
      continue;
    }

    improvedLines.push(line);
  }

  if (keywordsToAdd.length > addedKeywordIndex) {
    improvedLines.push("CORE KEYWORDS");
    improvedLines.push(keywordsToAdd.slice(addedKeywordIndex).join(" | "));
  }

  return {
    optimizedText: trimToOnePage(improvedLines.join("\n")),
    summary: `Optimized toward ${jobTitle} at ${companyName} with targeted JD keywords.`,
    addedKeywords: keywordsToAdd,
    improvedBullets: keywordsToAdd.slice(0, addedKeywordIndex),
    jobTitle,
    companyName
  };
}

function strengthenBullet(text: string, keyword: string) {
  const trimmed = text.trim();
  if (trimmed.toLowerCase().includes(keyword.toLowerCase())) {
    return trimmed;
  }

  return `${trimmed}; aligned work with ${keyword}`;
}

function getMissingKeywords(jd: string, resumeText: string) {
  const jdKeywords = extractKeywords(jd, 35);
  const resumeKeywords = new Set(extractKeywords(resumeText, 45));
  return jdKeywords.filter((keyword) => !resumeKeywords.has(keyword)).slice(0, 12);
}

export function trimToOnePage(text: string) {
  const words = text.replace(/\s+/g, " ").trim().split(" ");
  if (words.length <= 680) {
    return text.trim();
  }

  return `${words.slice(0, 680).join(" ")}\n\nAdditional older detail removed to keep this version to one page.`;
}

function clampScore(value: number) {
  return Math.max(0, Math.min(100, Math.round(value)));
}

File: app/optimizer.py
from __future__ import annotations

from typing import Any

from app.ai import call_openai_json
from app.keywords import extract_keywords, infer_job_details

OPTIMIZATION_SCHEMA = {
    "type": "object",
    "additionalProperties": False,
    "required": ["optimized_text", "summary", "added_keywords", "improved_bullets", "job_title", "company_name"],
    "properties": {
        "optimized_text": {"type": "string"},
        "summary": {"type": "string"},
        "added_keywords": {"type": "array", "items": {"type": "string"}},
        "improved_bullets": {"type": "array", "items": {"type": "string"}},
        "job_title": {"type": "string"},
        "company_name": {"type": "string"},
    },
}

ADVICE_SCHEMA = {
    "type": "object",
    "additionalProperties": False,
    "required": ["should_apply", "reasoning", "growth_potential_score", "skill_gap_analysis", "suggested_focus"],
    "properties": {
        "should_apply": {"type": "boolean"},
        "reasoning": {"type": "string"},
        "growth_potential_score": {"type": "number"},
        "skill_gap_analysis": {"type": "array", "items": {"type": "string"}},
        "suggested_focus": {"type": "array", "items": {"type": "string"}},
    },
}


def optimize_resume(resume: dict[str, Any], jd: str, match: dict[str, Any] | None, job_title: str, company_name: str) -> dict[str, Any]:
    details = infer_job_details(jd)
    title = job_title.strip() or details["job_title"]
    company = company_name.strip() or details["company_name"]
    missing_keywords = (match or {}).get("missing_keywords") or missing_keywords_for(jd, resume["text"])

    try:
        ai_result = call_openai_json(
            system=(
                "You are an expert ATS resume editor. Edit only resume text content. "
                "Keep the source resume section order and bullet style as much as plain text allows. "
                "Do not invent employers, credentials, dates, degrees, tools, or metrics. "
                "Keep the result concise enough for one page, around 450 to 650 words. Return JSON only."
            ),
            user={
                "job_title": title,
                "company_name": company,
                "job_description": jd,
                "current_resume_text": resume["text"],
                "missing_keywords": missing_keywords,
                "rules": [
                    "ATS-friendly",
                    "one page only",
                    "preserve section order",
                    "preserve bullet style",
                    "modify text only",
                ],
            },
            schema=OPTIMIZATION_SCHEMA,
            schema_name="optimized_resume",
        )
    except Exception:
        ai_result = None

    if ai_result and ai_result.get("optimized_text"):
        ai_result["optimized_text"] = trim_to_one_page(ai_result["optimized_text"])
        ai_result["job_title"] = title
        ai_result["company_name"] = company
        return ai_result

    return fallback_optimization(resume["text"], missing_keywords, title, company)


def career_advice(jd: str, resume: dict[str, Any] | None, match: dict[str, Any] | None) -> dict[str, Any]:
    resume_text = resume["text"] if resume else ""
    jd_keywords = extract_keywords(jd, 35)
    resume_keywords = extract_keywords(resume_text, 35)
    missing = [keyword for keyword in jd_keywords if keyword not in resume_keywords][:8]
    score = (match or {}).get("score", max(35, 86 - len(missing) * 7))

    try:
        ai_result = call_openai_json(
            system="You are a practical career advisor. Be direct, realistic, concise, and return JSON only.",
            user={
                "job_description": jd,
                "resume_text": resume_text,
                "match_score": score,
                "missing_keywords": missing,
            },
            schema=ADVICE_SCHEMA,
            schema_name="career_advice",
        )
    except Exception:
        ai_result = None

    if ai_result:
        ai_result["growth_potential_score"] = clamp_score(ai_result["growth_potential_score"])
        return ai_result

    return {
        "should_apply": score >= 58,
        "reasoning": (
            "The resume already overlaps with the JD on core responsibilities and keywords."
            if score >= 72
            else "The role is reachable with targeted resume optimization."
            if score >= 58
            else "The JD shows several gaps that may weaken the application."
        ),
        "growth_potential_score": clamp_score(score + min(12, len(missing) * 2)),
        "skill_gap_analysis": [f"Make {keyword} visible if you have real experience." for keyword in missing]
        or ["No major keyword gaps detected."],
        "suggested_focus": missing[:4],
    }


def fallback_optimization(text: str, missing_keywords: list[str], job_title: str, company_name: str) -> dict[str, Any]:
    lines = [line.rstrip() for line in text.splitlines() if line.strip()]
    keywords = missing_keywords[:8]
    keyword_index = 0
    improved: list[str] = []
    improved_bullets: list[str] = []

    for line in lines:
        stripped = line.strip()
        if stripped.startswith(("-", "*", "•")) and keyword_index < len(keywords) and len(stripped) < 180:
            keyword = keywords[keyword_index]
            keyword_index += 1
            new_line = f"{stripped}; aligned work with {keyword}"
            improved.append(new_line)
            improved_bullets.append(keyword)
        else:
            improved.append(line)

    if keyword_index < len(keywords):
        improved.append("CORE KEYWORDS")
        improved.append(" | ".join(keywords[keyword_index:]))

    return {
        "optimized_text": trim_to_one_page("\n".join(improved)),
        "summary": f"Optimized toward {job_title} at {company_name} with targeted JD language.",
        "added_keywords": keywords,
        "improved_bullets": improved_bullets,
        "job_title": job_title,
        "company_name": company_name,
    }


def missing_keywords_for(jd: str, resume_text: str) -> list[str]:
    resume_keywords = set(extract_keywords(resume_text, 45))
    return [keyword for keyword in extract_keywords(jd, 35) if keyword not in resume_keywords][:12]


def trim_to_one_page(text: str) -> str:
    words = text.replace("\n", " \n ").split()
    if len(words) <= 680:
        return text.strip()
    return " ".join(words[:680]).strip()


def clamp_score(value: float) -> int:
    return max(0, min(100, round(float(value))))
